crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs

```diff
@@ -0,0 +1,495 @@
+            SortedDunderAll::Sorted(value) => value,
+        };
+
+    let mut diagnostic = Diagnostic::new(UnsortedDunderAll, dunder_all_val.range());
+
+    if let Some(new_dunder_all) = new_dunder_all {
+        let applicability = {
+            if dunder_all_val.multiline
+                && checker.indexer().comment_ranges().intersects(node.range())
+            {
+                Applicability::Unsafe
+            } else {
+                Applicability::Safe
+            }
+        };
+        diagnostic.set_fix(Fix::applicable_edit(
+            Edit::range_replacement(new_dunder_all, dunder_all_val.range()),
+            applicability,
+        ));
+    }
+
+    checker.diagnostics.push(diagnostic);
+}
+
+struct DunderAllValue {
+    items: Vec<DunderAllItem>,
+    range: TextRange,
+    multiline: bool,
+}
+
+impl DunderAllValue {
+    fn from_expr(value: &ast::Expr, locator: &Locator) -> Option<DunderAllValue> {
+        // Step (1): inspect the AST to check that we're looking at something vaguely sane:
+        let is_multiline = locator.contains_line_break(value.range());
+        let (elts, range) = match value {
+            ast::Expr::List(ast::ExprList { elts, range, .. }) => (elts, range),
+            ast::Expr::Tuple(ast::ExprTuple { elts, range, .. }) => (elts, range),
+            _ => return None,
+        };
+        // An `__all__` definition with <2 elements can't be unsorted;
+        // no point in proceeding any further here
+        if elts.len() < 2 {
+            return None;
+        }
+        for elt in elts {
+            // Only consider sorting it if __all__ only has strings in it
+            let string_literal = elt.as_string_literal_expr()?;
+            // And if any strings are implicitly concatenated, don't bother
+            if string_literal.value.is_implicit_concatenated() {
+                return None;
+            }
+        }
+        // Step (2): parse the `__all__` definition using the raw tokens.
+        //
+        // (2a). Start by collecting information on each line individually:
+        let lines = collect_dunder_all_lines(*range, locator)?;
+
+        // (2b). Group lines together into sortable "items":
+        //   - Any "item" contains a single element of the `__all__` list/tuple
+        //   - "Items" are ordered according to the element they contain
+        //   - Assume that any comments on their own line are meant to be grouped
+        //     with the element immediately below them: if the element moves,
+        //     the comments above the element move with it.
+        //   - The same goes for any comments on the same line as an element:
+        //     if the element moves, the comment moves with it.
+        let items = collect_dunder_all_items(&lines);
+
+        Some(DunderAllValue {
+            items,
+            range: *range,
+            multiline: is_multiline,
+        })
+    }
+
+    fn construct_sorted_all(
```

Comment:
Nit: I think idiomatically in Rust you might call this `as_sorted()` (if it takes `&self`) or `into_sorted()`. In this case, I think `into_sorted` would make sense, then it could take `self` instead of `&self` and consume `self.items` (so no need to `.clone()`).